tools/ld-dropout-correct/dropoutcorrect.cpp

```diff
@@ -294,83 +327,124 @@ DropOutCorrect::Replacement DropOutCorrect::findReplacementLine(const QVector<Dr
+                        candidate.fieldLine << "distance" << distance << "of source" << candidate.sourceNumber <<
+                        "with a quality of" << candidate.quality;
 
+            // Choose the candidate if the distance is less
             if (distance < lowestDistance) {
                 replacement = candidate;
+                replacement.distance = distance;
+                lowestDistance = distance;
+            }
+
+            // Choose the candidate if the distance is the same, but the quality is better
+            else if (distance == lowestDistance && candidate.quality > replacement.quality) {
+                replacement = candidate;
+                replacement.distance = distance;
                 lowestDistance = distance;
             }
         }
     }
 
+    if (replacement.fieldLine != -1) {
+        qDebug() << "Selected replacement is" <<
+                    (replacement.isSameField ? "same" : "other") << "field, line" <<
+                    replacement.fieldLine << "of source" << replacement.sourceNumber << (matchChromaPhase ? "(chroma phase matched)" : "(whole signal)") <<
+                    "with a quality of" << replacement.quality;
+    } else {
+        qDebug() << "No viable replacement selected for" << thisFieldDropouts[0][dropOutIndex].fieldLine;
+    }
+
+
     return replacement;
 }
 
 // Given a dropout, scan through a source field for the nearest replacement line that doesn't have overlapping dropouts.
 // Adds a Replacement to candidates if one was found.
-void DropOutCorrect::findPotentialReplacementLine(const QVector<DropOutLocation> &targetDropouts, qint32 targetIndex,
-                                                  const QVector<DropOutLocation> &sourceDropouts, bool isSameField,
+void DropOutCorrect::findPotentialReplacementLine(const QVector<QVector<DropOutLocation>> &targetDropouts, qint32 targetIndex,
+                                                  const QVector<QVector<DropOutLocation>> &sourceDropouts, bool isSameField,
                                                   qint32 sourceOffset, qint32 stepAmount,
                                                   qint32 firstActiveFieldLine, qint32 lastActiveFieldLine,
-                                                  QVector<Replacement> &candidates)
-{
-    qint32 sourceLine = targetDropouts[targetIndex].fieldLine + sourceOffset;
+                                                  QVector<Replacement> &candidates, qint32 sourceNo,
+                                                  QVector<qreal> sourceFrameQuality)
+{    
+    // Calculate the start source line (which is the same line as the dropout unless the source number is 0
+    qint32 sourceLine = targetDropouts[0][targetIndex].fieldLine;
+    if (sourceNo == 0) sourceLine += sourceOffset;
+
+    // Is the line within the active range?
+    if (sourceLine < firstActiveFieldLine || sourceLine > lastActiveFieldLine) {
```

Comment:
`>=` rather than `>`? (For consistency with `< lastActiveFieldLine` below.)

Reply:
Done